plugins/optimization-detective/class-od-tag-visitor-registry.php

```diff
@@ -35,11 +35,21 @@ final class OD_Tag_Visitor_Registry implements Countable, IteratorAggregate {
 	 *
 	 * @phpstan-param TagVisitorCallback $tag_visitor_callback
 	 *
-	 * @param string   $id                   Identifier for the tag visitor.
+	 * @param string   $id Identifier for the tag visitor.
 	 * @param callable $tag_visitor_callback Tag visitor callback.
+	 * @param string[] $dependencies Tag visitors that must run before this tag visitor.
 	 */
-	public function register( string $id, callable $tag_visitor_callback ): void {
-		$this->visitors[ $id ] = $tag_visitor_callback;
+	public function register( string $id, callable $tag_visitor_callback, array $dependencies = array() ): void {
+		if ( ! empty( $dependencies ) ) {
+			foreach ( $dependencies as $dependency_id => $dependency_callback ) {
+				if ( ! self::is_registered( $dependency_id ) ) {
+					$this->register( $dependency_id, $dependency_callback );
+				}
+			}
+		}
+		if ( ! self::is_registered( $id ) ) {
+			$this->visitors[ $id ] = $tag_visitor_callback;
+		}
```

Comment:
I think this should go back to how it was before, but refactor the `visitors` member variable so that it is not just a mapping of ID to callable, but rather a mapping of ID to an array of information about the callable, for example:

```suggestion
		$this->visitors[ $id ] = array(
			'callback'     => $tag_visitor_callback,
			'dependencies' => $dependencies,
		);
```

Then, instead of implementing the dependency handling here, it can be implemented in the `getIterator()` method:

https://github.com/WordPress/performance/blob/16b896d533fa865989f481333041669ab46b8986/plugins/optimization-detective/class-od-tag-visitor-registry.php#L92-L99

This method would end up looking a lot like `WP_Scripts::all_deps()` in core (but much much simpler) which recursively goes through the dependencies to ensure the right order in the end.

If there is a tag visitor that lacks a dependency (recursively), then it would get omitted in the same way as scripts get omitted when one of their dependencies is not registered. It may be appropriate to include a `_doing_it_wrong()` in this case as well.